### PR TITLE
Improved `avoid_late_keyword` to support ignoring the subtype of the node type (#157)

### PR DESCRIPTION
**Pull request by Yarl745**
* Fixed issue #157

### CODE REVIEW

File: lint_test/avoid_late_keyword_test.dart [hunk 1]
```diff
@@ -37,6 +39,20 @@ class AvoidLateKeyword {
...
   late final na2 = NotAllowed();
 
+  /// expect_lint: avoid_late_keyword
+  late final Subscription<String> subscription1;
+
+  late final Subscription<List<String>> subscription2;
+
+  late final Subscription<List<List<int>>> subscription3;
+
+  late final Subscription<Map<dynamic, String>> subscription4;
```
Comment (yurii-prykhodko-solid) [inline lint_test/avoid_late_keyword_test.dart:49]:
```suggestion
  late final Subscription<Map<String, String>> subscription7; // no lint
```

File: lib/src/utils/types_utils.dart [hunk 2]
```diff
@@ -32,6 +32,95 @@ extension Subtypes on DartType {
...
+    return withGenerics ? displayString : displayString.replaceGenericString();
+  }
+
+  /// Checks if a variable type is among the ignored types.
+  bool hasIgnoredType({required Set<String> ignoredTypes}) {
+    if (ignoredTypes.isEmpty) return false;
+
+    final checkedTypes = [this, ...supertypes];
+
+    ignoredTypes = getIntersectionTypesFor(ignoredTypes: ignoredTypes);
```
Comment (yurii-prykhodko-solid) [inline lib/src/utils/types_utils.dart:52]: Don't mutate argument variables.

```suggestion
    final intersectionIgnoredTypes = getIntersectionTypesFor(ignoredTypes: ignoredTypes);
```

File: lib/src/utils/types_utils.dart [hunk 3]
```diff
@@ -32,6 +32,95 @@ extension Subtypes on DartType {
...
+  String replaceGenericString() => replaceFirst(_genericRegex, '');
+
+  RegExp getGeneralTypeRegex() {
+    var out = this;
+    for (final regex in _baseTypesRegex) {
+      out = out.replaceAll(regex, _baseTypeReplacement);
+    }
+
+    return RegExp(out);
+  }
```
Comment (yurii-prykhodko-solid) [inline lib/src/utils/types_utils.dart:123]: Nitpick: you can use `fold` or `reduce` over _baseTypesRegex to shorten this.

File: lib/src/utils/types_utils.dart [hunk 4]
```diff
@@ -32,6 +32,95 @@ extension Subtypes on DartType {
...
+        uniqueCheckedTypeStrings.intersection(uniqueIgnoredTypeStrings);
+
+    // Filters and returns the set of ignored types that match
+    // any intersection types.
+    return ignoredTypes.where(
+      (ignoredType) {
+        return intersectionUniqueTypes.firstWhereOrNull(
+              (uniqueType) => ignoredType.contains(uniqueType),
+            ) !=
+            null;
```
Comment (yurii-prykhodko-solid) [inline lib/src/utils/types_utils.dart:97]: `any` instead of `firstWhereOrNull(...) != null`?

**Review by yurii-prykhodko-solid — CHANGES_REQUESTED** [review]
General logic LGTM, great start!;
Implementation-wise -- I think we are relying on raw regexp/string matching too much here. We should try and leverage the static types and utilities that analyzer/ast provides us.

Here's some stuff we may be able to use:
- https://pub.dev/documentation/analyzer/latest/dart_ast_token/Keyword/DYNAMIC-constant.html
- https://pub.dev/documentation/analyzer/latest/dart_element_type/DynamicType-class.html
- https://pub.dev/documentation/analyzer/latest/dart_element_type_provider/TypeProvider/objectType.html
- https://pub.dev/documentation/analyzer/latest/dart_element_type/TypeParameterType-class.html

We should also be more vigorous about testing this. I've added a couple ideas, but I'm sure you can expand on them.

File: lib/src/utils/node_utils.dart
Comment (Yarl745) [inline lib/src/utils/node_utils.dart:1]: - Enhances the AST analysis capabilities for type-related nodes by introducing a `TypeAnalyzerNode` class.
- Extends `GeneralizingAstVisitor` to navigate and process AST nodes, focusing on type analysis.
- Includes functionalities such as identifying 'dynamic' and 'Object' types.
- Manages child nodes and incorporates factory constructors for initialization from strings or AST nodes.
- Provides utility methods like `isInclude` for node comparison.
- Integrates visitor patterns to handle named types effectively.

File: lib/src/utils/node_utils.dart [hunk 5]
```diff
@@ -22,3 +24,88 @@ String humanReadableNodeType(AstNode? node) {
...
+    final analyzerNode = TypeAnalyzerNode();
+    node.visitChildren(analyzerNode);
+    return analyzerNode;
+  }
+
+  /// Determines if the current node includes another analyzer node.
+  bool isInclude({required TypeAnalyzerNode node}) {
+    if (isDynamicType || isObjectType) return true;
+
+    if (this != node) return false;
```
Comment (Yarl745) [inline lib/src/utils/node_utils.dart:72]: Perhaps need to change the name of the method to `isSubtype`

File: lib/src/lints/avoid_late_keyword/avoid_late_keyword_rule.dart [hunk 6]
```diff
@@ -96,16 +96,11 @@ class AvoidLateKeywordRule extends SolidLintRule<AvoidLateKeywordParameters> {
   }
 
   bool _hasIgnoredType(VariableDeclaration node) {
-    final ignoredTypes = config.parameters.ignoredTypes.toSet();
-    if (ignoredTypes.isEmpty) return false;
-
```
Comment (yurii-prykhodko-solid) [inline lib/src/lints/avoid_late_keyword/avoid_late_keyword_rule.dart:101]: We should probably keep this.

File: lib/src/utils/node_utils.dart [hunk 7]
```diff
@@ -22,3 +24,66 @@ String humanReadableNodeType(AstNode? node) {
...
+class _NamedTypeFinder extends GeneralizingAstVisitor<void> {
+  NamedType? _foundNamedType;
+
+  NamedType? get foundNamedType => _foundNamedType;
+
+  @override
+  void visitNamedType(NamedType namedType) {
+    _foundNamedType ??= namedType;
+  }
+}
```
Comment (yurii-prykhodko-solid) [inline lib/src/utils/node_utils.dart:53]: Let's place this in a separate file, to limit `_NamedTypeFinder` visibility scope.

File: lib/src/utils/node_utils.dart [hunk 8]
```diff
@@ -22,3 +24,66 @@ String humanReadableNodeType(AstNode? node) {
...
+extension ChildNamedTypes on NamedType {
+  /// Retrieves child [NamedType] instances from type arguments.
+  List<NamedType> get childNamedTypes =>
+      typeArguments?.arguments.whereType<NamedType>().toList() ?? [];
+
+  /// Gets the token name of this type instance.
+  String get tokenName => name2.toString();
+
+  /// Checks if the current token name is 'dynamic'.
+  bool get isDynamic => tokenName == "dynamic";
```
Comment (yurii-prykhodko-solid) [inline lib/src/utils/node_utils.dart:65]: ?
```suggestion
  bool get isDynamic => this is DynamicType;
```

File: lib/src/utils/node_utils.dart [hunk 9]
```diff
@@ -22,3 +24,66 @@ String humanReadableNodeType(AstNode? node) {
...
+
+  /// Checks if the current token name is 'Object'.
+  bool get isObject => tokenName == "Object";
```
Comment (yurii-prykhodko-solid) [inline lib/src/utils/node_utils.dart:68]: ?
```suggestion
  /// Checks if the current token name is 'Object'.
  bool get isObject => type.isDartCoreObject;
```

File: lib/src/utils/types_utils.dart [hunk 10]
```diff
@@ -22,16 +22,67 @@
...
+    }
+
+    return false;
+  }
+}
+
+extension TypeString on String {
+  static final _genericRegex = RegExp('<.*>');
+
+  bool get hasGenericString => contains(_genericRegex);
```
Comment (yurii-prykhodko-solid) [inline lib/src/utils/types_utils.dart:83]: seems unused
```suggestion
```

File: lib/src/utils/node_utils.dart [hunk 11]
```diff
@@ -22,3 +24,66 @@ String humanReadableNodeType(AstNode? node) {
...
+
+  /// Checks if this node is a subtype of the specified node
+  /// based on their structures.
+  bool isSubtype({required NamedType node}) {
```
Comment (yurii-prykhodko-solid) [inline lib/src/utils/node_utils.dart:72]: I think this name communicates more clearly the direction of the comparison.

```suggestion
  bool isSubtypeOf({required NamedType node}) {
```

File: lint_test/avoid_late_keyword_allow_no_type_specified/avoid_late_keyword_allow_no_type_specified.dart [hunk 12]
```diff
@@ -0,0 +1,34 @@
+// ignore_for_file: prefer_const_declarations, unused_local_variable, prefer_match_file_name
```
Comment (yurii-prykhodko-solid) [inline lint_test/avoid_late_keyword_allow_no_type_specified/avoid_late_keyword_allow_no_type_specified.dart:1]: move this to lint_test/avoid_late_keyword/no_generics_specified

File: lint_test/avoid_late_keyword_test.dart [hunk 13]
```diff
@@ -9,6 +9,10 @@ class SubAnimationController extends AnimationController {}
```
Comment (yurii-prykhodko-solid) [inline lint_test/avoid_late_keyword_test.dart:9]: lint_test/avoid_late_keyword/with_generics

**Review by yurii-prykhodko-solid — CHANGES_REQUESTED** [review]

**Review by yurii-prykhodko-solid — APPROVED** [review]
LGTM!

**Review by illia-romanenko — APPROVED** [review]
LGTM - thanks

**Pull request merged**